### ISSUE TITLE
`torch.roll` with `fill_` produces wrong outputs for specific input shapes

### ISSUE THREAD

**Issue by ptrblck** [pytorch/pytorch#117778]
### 🐛 Describe the bug

Minimal code snippet to reproduce the issue:
```python
import torch

#torch.set_printoptions(threshold=10000)

def roll_and_zero_first_row(x):
    x = torch.roll(x, -1, 0)
    x[0].fill_(0.0)
    return x

size = (1024, 2)

inp = torch.ones(size, device="cuda")
compiled = torch.compile(roll_and_zero_first_row)
out_eager = roll_and_zero_first_row(inp)
out_compiled = compiled(inp)

print((out_eager - out_compiled).abs())
print(out_eager)
print(torch.testing.assert_allclose(out_eager, out_compiled))
```
Comment the `printoptions` in if you want to see the actual mismatches.

Originally reported here: https://github.com/NVIDIA/TransformerEngine/issues/600

Thanks @Marks101 for the great debugging and for providing the code snippet.

Attaching the generated code from `TORCH_LOGS=output_code`.
[triton_roll.txt](https://github.com/pytorch/pytorch/files/13981522/triton_roll.txt)

### Versions

Reproduced on an A100 with `2.1.2+cu121` as well as the latest nightly `pytorch-triton-2.2.0+e28a256d71` `torch-2.3.0.dev20240118+cu121`.

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519

**Comment by Marks101** [comment]
One additional note maybe: i already opend https://github.com/openai/triton/issues/2941 because I suspect the issue to arise from triton .

Thanks @ptrblck for reproducing this and forwarding it!

**Comment by ptrblck** [comment]
Great! We are also checking if `ptxas` creates the expected output or might cause the issue.

**Comment by yf225** [comment]
Update: this seems a `ptxas` bug according to https://github.com/openai/triton/issues/2941#issuecomment-1917495713, and it's fixed in CUDA 12.4
